### PR TITLE
[Ebook] Ebook words list

### PR DESCRIPTION
**Pull request by kastrubait**

### CODE REVIEW

File: rslang/src/app/auth/components/login/login.component.html [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<section class="auth-page">
+  <div class="auth-page__wrapper">
+    <mat-card>
+      <form [formGroup]="loginForm" (ngSubmit)="submitForm()">
+        <mat-form-field>
+          <input matInput id="email" type="email" placeholder="Email" formControlName="email" required />
+          <div *ngIf="emailInput && emailInput.invalid && (emailInput.dirty || emailInput.touched)">
```
Comment (Nikitzu) [inline rslang/src/app/auth/components/login/login.component.html]: I thnik we can use only mat-error without this div

File: rslang/src/app/ebook/components/ebook-header/ebook-header.component.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  templateUrl: './ebook-header.component.html',
+  styleUrls: ['./ebook-header.component.scss'],
+})
+export class EbookHeader {
+  totalPages = 30;
+  title = 'ebook';
+  @Output() pageNoChanged: EventEmitter<number> = new EventEmitter<number>();
+
+  onPageChange(pageNo: number): void {
+    this.pageNoChanged.emit(pageNo);
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/ebook-header/ebook-header.component.ts]: better pageChanged

File: rslang/src/app/ebook/components/ebook-home/ebook-home.component.ts [hunk 3]
```diff
@@ -52,6 +49,18 @@ export class EbookHome implements OnInit, OnDestroy {
       .setItem(LocalStorageKey.EbookSettings, JSON.stringify(this.userBookSettings));
   }
 
+  changeSelectedBookPage(pageNoChanged: number): void {
+    this.userBookSettings.currentState.page = pageNoChanged;
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/ebook-home/ebook-home.component.ts]: same

File: rslang/src/app/ebook/components/words-list/words-list.component.html [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<app-ebook-header (pageNoChanged)="changeSelectedPage($event)"> </app-ebook-header>
+<section class="words-conteiner">
+  <ng-container *ngIf="words.length > 0; else isEmptyCart" class="words-conteiner__list">
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/words-list/words-list.component.html]: !!words.length

File: rslang/src/app/ebook/components/words-list/words-list.component.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import { LocalStorageService } from '../../../core/services/local-storage.service';
+
+@Component({
+  selector: 'app-words-list',
+  templateUrl: './words-list.component.html',
+  styleUrls: ['./words-list.component.scss'],
+})
+export class WordsList implements OnInit {
+  private subscriptions: Subscription[] = [];
+  set subscription(sb: Subscription) { this.subscriptions.push(sb); }
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/words-list/words-list.component.ts:28]: why there is no unsubscribe?

File: rslang/src/app/ebook/directives/pagination.directive.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+      this.setValue(this.totalPages);
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    this.pageNo = Number(this.el.nativeElement.value);
+    this.pageChange.emit(this.pageNo);
+  }
+
+  get isFirst(): boolean {
+    return this.pageNo === 1;
```
Comment (Nikitzu) [inline rslang/src/app/ebook/directives/pagination.directive.ts]: pageNumber

File: rslang/src/app/ebook/directives/pagination.directive.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  get isLast(): boolean {
+    return this.pageNo === this.totalPages;
```
Comment (Nikitzu) [inline rslang/src/app/ebook/directives/pagination.directive.ts]: same

File: rslang/src/app/shared/services/words-data.service.ts [hunk 8]
```diff
@@ -11,7 +14,26 @@ export class WordsDataService extends BaseDataService<Word[]> {
...
   }
 
-  public GetWords(): Word[] {
+  GetWords(option: CurrentStateBook): Word[] {
+    const apiWords: Word[] = [];
+    const params: HttpParams = new HttpParams()
+      .set('group', String(option.group - 1))
+      .set('page', String(option.page - 1));
+    this.getData(API_URL.WORDS, { params });
+    this.data$.subscribe((words: Word[]) => {
```
Comment (Nikitzu) [inline rslang/src/app/shared/services/words-data.service.ts]: why dataService is subscribed to its own data stream??

**Review by Nikitzu — CHANGES_REQUESTED** [review]

File: rslang/src/app/ebook/directives/pagination.directive.ts (hunk 6 above)
Comment (kastrubait) [reply rslang/src/app/ebook/directives/pagination.directive.ts]: fix

File: rslang/src/app/ebook/directives/pagination.directive.ts (hunk 7 above)
Comment (kastrubait) [reply rslang/src/app/ebook/directives/pagination.directive.ts]: fix

File: rslang/src/app/ebook/components/ebook-header/ebook-header.component.ts (hunk 2 above)
Comment (kastrubait) [reply rslang/src/app/ebook/components/ebook-header/ebook-header.component.ts]: fix

File: rslang/src/app/ebook/components/words-list/words-list.component.html (hunk 4 above)
Comment (kastrubait) [reply rslang/src/app/ebook/components/words-list/words-list.component.html]: fix

File: rslang/src/app/ebook/components/ebook-header/ebook-header.component.html [hunk 9]
```diff
@@ -0,0 +1,61 @@
+<section class="ebook-header">
+    <div class="ebook-header__title">
+        <h1>{{ title }}</h1>
+    </div>
+    <div class="ebook-header__pagination">
+        <button (click)="pagination.first()" [disabled]="pagination.isFirst"><<</button>
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/ebook-header/ebook-header.component.html:6]: what is pagination - I dont see such property in ts

File: rslang/src/app/ebook/components/word-item/word-item.component.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  selector: 'app-word-item',
+  templateUrl: './word-item.component.html',
+  styleUrls: ['./word-item.component.scss'],
+})
+export class WordItem {
+  @Input() item: Word;
+  @Output() wordDifficult: EventEmitter<string> = new EventEmitter<string>();
+  @Output() wordRemove: EventEmitter<string> = new EventEmitter<string>();
+
+  turnOverItem(): string {
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/word-item/word-item.component.ts:18]: We really need a method for this static string?

File: rslang/src/app/ebook/components/words-list/words-list.component.ts [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  changeDifficultWord(wordDifficult: string): void {
+    console.log(this.words[this.indexWord(wordDifficult)]);
+  }
+
+  indexWord(wordId: string): number {
+    return this.words.findIndex((element: Word) => element.id === wordId);
+  }
+
+  getWordsList(currentState: CurrentStateBook): void {
+    this.words = [];
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/words-list/words-list.component.ts:124]: This whole method needs to be changed

File: rslang/src/app/ebook/components/words-list/words-list.component.ts [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+  changeRemoveWord(wordRemove: string): void {
+    console.log(this.words[this.indexWord(wordRemove)]);
+  }
+
+  changeDifficultWord(wordDifficult: string): void {
+    console.log(this.words[this.indexWord(wordDifficult)]);
+  }
+
+  indexWord(wordId: string): number {
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/words-list/words-list.component.ts]: wordIndex

File: rslang/src/app/ebook/components/words-list/words-list.component.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+      .setItem(LocalStorageKey.EbookSettings, JSON.stringify(this.userBookSettings));
+    const { currentState } = this.userBookSettings;
+    this.getWordsList(currentState);
+  }
+
+  changeRemoveWord(wordRemove: string): void {
+    console.log(this.words[this.indexWord(wordRemove)]);
+  }
+
+  changeDifficultWord(wordDifficult: string): void {
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/words-list/words-list.component.ts]: changeWordDifficulty?

File: rslang/src/app/ebook/components/words-list/words-list.component.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+  changeSelectedPage(pageNoChanged: number): void {
+    this.userBookSettings.currentState.page = pageNoChanged;
+    this.localStorageService
+      .setItem(LocalStorageKey.EbookSettings, JSON.stringify(this.userBookSettings));
+    const { currentState } = this.userBookSettings;
+    this.getWordsList(currentState);
+  }
+
+  changeRemoveWord(wordRemove: string): void {
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/words-list/words-list.component.ts]: changeRemove? so change or remove?

File: rslang/src/app/ebook/directives/pagination.directive.ts [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+  get isFirst(): boolean {
+    return this.pageNumber === 1;
+  }
+
+  get isLast(): boolean {
+    return this.pageNumber === this.totalPages;
+  }
+
+  first(): void {
```
Comment (Nikitzu) [inline rslang/src/app/ebook/directives/pagination.directive.ts:56]: Why we need methoid for such action? Cant we call setPage(1) from page?

File: rslang/src/app/ebook/directives/pagination.directive.ts [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+  prev(): void {
+    this.setPage(Math.max(1, this.pageNumber - 1));
+  }
+
+  next():void {
+    this.setPage(Math.min(this.totalPages, this.pageNumber + 1));
+  }
+
+  last(): void {
+    this.setPage(this.totalPages);
```
Comment (Nikitzu) [inline rslang/src/app/ebook/directives/pagination.directive.ts:69]: same here

File: rslang/src/app/ebook/directives/pagination.directive.ts [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+    return Number(val) > this.totalPages;
+  }
+
+  private onTotalPagesInput() {
+    if (typeof this.totalPages !== 'number') {
+      this.totalPages = 1;
+    }
+  }
+
+  private onpageNumberInput() {
```
Comment (Nikitzu) [inline rslang/src/app/ebook/directives/pagination.directive.ts:96]: onPageNumberInput

File: rslang/src/app/ebook/directives/pagination.directive.ts [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+    if (
+      typeof this.pageNumber !== 'number'
```
Comment (Nikitzu) [inline rslang/src/app/ebook/directives/pagination.directive.ts:98]: isn't it always true?

File: rslang/src/app/shared/services/words-data.service.ts [hunk 19]
```diff
@@ -11,7 +15,15 @@ export class WordsDataService extends BaseDataService<Word[]> {
     super(httpClient);
   }
 
-  public GetWords(): Word[] {
+  GetWords(option: CurrentStateBook): Observable<Word[]> {
+    const params: HttpParams = new HttpParams()
+      .set('group', String(option.group - 1))
+      .set('page', String(option.page - 1));
+    this.getData(API_URL.WORDS, { params });
+    return this.data$;
```
Comment (Nikitzu) [inline rslang/src/app/shared/services/words-data.service.ts]: all should be refactored

**Review by Nikitzu — CHANGES_REQUESTED** [review]

File: rslang/src/app/shared/services/words-data.service.ts (hunk 19 above)
Comment (kastrubait) [reply rslang/src/app/shared/services/words-data.service.ts]: was refactored

File: rslang/src/app/ebook/components/words-list/words-list.component.ts (hunk 12 above)
Comment (kastrubait) [reply rslang/src/app/ebook/components/words-list/words-list.component.ts]: why?

File: rslang/src/app/ebook/components/words-list/words-list.component.ts (hunk 11 above)
Comment (kastrubait) [reply rslang/src/app/ebook/components/words-list/words-list.component.ts:124]: changed

File: rslang/src/app/ebook/components/word-item/word-item.component.ts (hunk 10 above)
Comment (kastrubait) [reply rslang/src/app/ebook/components/word-item/word-item.component.ts:18]: Yes! This is a very important part of the idea

File: rslang/src/app/ebook/components/ebook-header/ebook-header.component.html (hunk 9 above)
Comment (kastrubait) [reply rslang/src/app/ebook/components/ebook-header/ebook-header.component.html:6]: this is not a property, but a directive

File: rslang/src/app/ebook/components/ebook-home/ebook-home.component.ts (hunk 3 above)
Comment (kastrubait) [reply rslang/src/app/ebook/components/ebook-home/ebook-home.component.ts]: fix

File: rslang/src/app/ebook/directives/pagination.directive.ts (hunk 15 above)
Comment (kastrubait) [reply rslang/src/app/ebook/directives/pagination.directive.ts:56]: I suppose it's possible, but I want to make these corrections later. Is it possible?

File: rslang/src/app/ebook/directives/pagination.directive.ts (hunk 16 above)
Comment (kastrubait) [reply rslang/src/app/ebook/directives/pagination.directive.ts:69]: I suppose it's possible, but I want to make these corrections later. Is it possible?

File: rslang/src/app/ebook/components/word-item/word-item.component.ts (hunk 10 above)
Comment (Nikitzu) [reply rslang/src/app/ebook/components/word-item/word-item.component.ts:18]: Explain please. This method is called by click. Why it has return? Where we return this value and why?

File: rslang/src/app/ebook/components/ebook-home/ebook-home.component.ts (hunk 3 above)
Comment (Nikitzu) [reply rslang/src/app/ebook/components/ebook-home/ebook-home.component.ts]: pageNoChanged

File: rslang/src/app/ebook/components/words-list/words-list.component.html [hunk 20]
```diff
@@ -0,0 +1,15 @@
+<app-ebook-header (pageChanged)="changeSelectedPage($event)"> </app-ebook-header>
+<section class = "words-conteiner">
+  <ng-container *ngIf = "!!words.length; else isEmptyCart" class="words-conteiner">
+    <app-word-item
+      class = "words-conteiner__item"
+      *ngFor = "let item of words | filterByAction: ''"
+      [item] = "item"
+      (setAction)=setActionForWord($event)>
+    </app-word-item>
+  </ng-container>
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/words-list/words-list.component.html:10]: in this file for some reason all '=' signs are surrounded by spaces.

File: rslang/src/app/ebook/components/words-list/words-list.component.spec.ts [hunk 21]
```diff
@@ -2,14 +2,15 @@ import { ComponentFixture, TestBed } from '@angular/core/testing';
...
-describe('ListWords', () => {
+describe('WordsListComponent', () => {
   let component: WordsList;
   let fixture: ComponentFixture<WordsList>;
 
   beforeEach(async () => {
     await TestBed.configureTestingModule({
-      declarations: [WordsList],
-    }).compileComponents();
+      declarations: [ WordsList ]
```
Comment (Nikitzu) [inline rslang/src/app/ebook/components/words-list/words-list.component.spec.ts]: spaces

File: rslang/src/app/shared/services/user-word-action.service.ts [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+// import { NotificationService } from '../../shared/services/notification.service';
+import { CallbackObject } from '../models/callback-object.model';
+import { CreateUserWordRequest } from '../models/create-user-word-request.model';
+import { API_URL } from '../constants/api-url';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class UserWordActionService extends BaseActionService {
+  // eslint-disable-next-line @typescript-eslint/no-useless-constructor
```
Comment (Nikitzu) [inline rslang/src/app/shared/services/user-word-action.service.ts]: no such comments

**Review by Nikitzu — CHANGES_REQUESTED** [review]

File: rslang/src/app/ebook/components/words-list/words-list.component.spec.ts (hunk 21 above)
Comment (kastrubait) [reply rslang/src/app/ebook/components/words-list/words-list.component.spec.ts]: fix

**Review by Nikitzu — APPROVED** [review]

**Review by coriander31415 — APPROVED** [review]

**Pull request merged**